SO-4640: Change Date to String in create version API

cmark opened the pull request:
Change `effectiveTime` property to String type to avoid possible timezone issues on Java API level.
This is an improvement on the existing functionality, everything should behave the same as before.

cmark committed:
SO-4640 #resolve #time 30m

Change `effectiveTime` property to String type to avoid timezone issues
on Java API level.

cmark committed:
SO-4640: remove unnecessary DatastoreOperationLockException

Use OperationLockException instead.
Simplify lock exception handling.

cmark committed:
SO-4640: remove unused Dates.todayLocal()

cmark committed:
SO-4640: access User from context instead of from RemoteJob

cmark committed:
SO-4640: fix DelegatingContext.getBindings issue

Current context bindings should override delegate context's bindings and
not the other way around.

cmark committed:
SO-4640: make sure we always get the timestamp from the start of the day

...in GMT timezone `EffectiveTimes.getEffectiveTime(Date)`

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/codesystem/CodeSystemVersionProperties.java:
```diff
@@ -47,6 +45,6 @@
 	 * @return the effective date of this code system version (can be {@code null})
 	 */
 	@JsonFormat(shape=Shape.STRING, pattern="yyyyMMdd")
```
Does this annotation do anything with a `String` typed property?

cmark replied:
Yes, I'm currently testing this, will let you know once I have something, I hope it does, if not, I'll remove it and add a default value to the API. Also, the `LocalDate` type could also work on the REST API.

cmark replied:
Cool, it renders the same example model in Swagger:

```
{
  "description": "string",
  "effectiveDate": "yyyyMMdd",
  "version": "string"
}
```

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/codesystem/version/CodeSystemVersionCreateRequest.java:
```diff
@@ -92,11 +90,11 @@
 	public Boolean execute(ServiceProvider context) {
-		final RemoteJob job = context.service(RemoteJob.class);
-		final String user = job.getUser();
+		final String user = context.service(User.class).getUsername();
```
Do we have the `User` injected for requests that are running in a remote job?

cmark replied:
Yes, otherwise this would fail.

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/codesystem/version/CodeSystemVersionCreateRequest.java:
```diff
@@ -221,13 +221,6 @@ private void acquireLocks(ServiceProvider context, String user, Collection<CodeS
 				lockTargetsByContext.put(lockContext, lockTarget);
 			}
-			
-		} catch (final OperationLockException e) {
```
Is this handled gracefully somewhere else?

cmark replied:
Hmm, this needs to be converted to `LockedException`, I'll add something to make that easier. Or we could throw that exception instead of any low-level ones.

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/codesystem/version/VersioningRequest.java:
```diff
@@ -99,7 +99,7 @@ private final CodeSystemVersionEntry createVersion(final TransactionContext cont
 		return CodeSystemVersionEntry.builder()
```
⚠️ This file also needs a copyright notice update

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/date/EffectiveTimes.java:
```diff
@@ -157,7 +156,12 @@ public static final long getEffectiveTime(@Nullable Date effectiveTime) {
 		}
-		return effectiveTime.getTime();
+		return effectiveTime.toInstant()
+			      .atZone(ZoneId.of("GMT"))
```
For consistency:

```suggestion
			      .atZone(ZoneOffset.UTC)
```

cmark replied:
Actually, this is how it looked like initially, but I intentionally changed it, since one requires a ZoneId and the other requires a ZoneOffset, anyway, the logic will remain the same, so happy to change it back. 😄

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/repository/RepositoryTransactionContext.java:
```diff
@@ -277,8 +277,8 @@ public Long commit(String author, String commitComment, String parentLockContext
 		final DatastoreLockTarget lockTarget = createLockTarget(id(), path());
 		IOperationLockManager locks = service(IOperationLockManager.class);
 		Commit commit = null;
+		acquireLock(locks, lockContext, lockTarget);
```
I don't remember how the locking goes internally, but if some locks can be granted when another fails, the granted ones should be released. In this case, `acquireLock` should remain inside the `try-finally` block.

cmark replied:
Hmm, right, then we need to omit/skip locks that cannot be released. 😕

apeteri reviewed (approved):
👍 I left some questions below.

cmark committed:
SO-4640: move common CodeSystem and Version related REST API calls...

...to test.commons

cmark committed:
Update core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/date/EffectiveTimes.java

Co-authored-by: András Péteri <apeteri@b2international.com>

cmark committed:
SO-4640: copyright notice update

cmark committed:
SO-4640: replace OperationLockException with LockedException

Also, remove InterruptedException from lock method signature (handle
internally by throwing a runtime error).

cmark committed:
SO-4640: rename DatastoreOperationLockManager...

...to DefaultOperationLockManager

cmark committed:
SO-4640: copyright notice update

cmark committed:
SO-4640: use more descriptive locked error message

Do NOT throw LockedException when trying to release a not locked
context. Report warning log message instead.

cmark committed:
SO-4640: add filterByEffectiveDate(long, long) method to version search

Deprecate Date based effective date filter method.
